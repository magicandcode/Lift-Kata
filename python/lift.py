from enum import Enum

from attr import dataclass


class Direction(Enum):
    UP = 0
    DOWN = 1


@dataclass
class Call:
    floor: int
    direction: Direction


class Lift:

    def __init__(self, id, floor, doors_open=False, requested_floors=None):
        self.id = id
        self.floor = floor
        self.doors_open = doors_open
        self.requested_floors = set(requested_floors) if requested_floors else []

    def request_floor(self, floor):
        self.requested_floors.append(floor)

    def tick(self):
        if self.floor in self.requested_floors:
            self.requested_floors = set()
            self.doors_open = True
        else:
            self.floor += 1


class LiftSystem:
    def __init__(self, floors=None, calls=None, lifts=None):
        self.floors = list(floors) if floors else []
        self.calls = list(calls) if calls else []
        self.lifts = list(lifts) if lifts else []

    def calls_for(self, floor):
        return [c for c in self.calls if c.floor == floor]

    def tick(self):
        for lift in self.lifts:
            lift.tick()



