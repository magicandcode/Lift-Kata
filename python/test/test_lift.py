from approvaltests import verify

from lift import Lift, LiftSystem
from lift_printers import print_lifts


def test_lift_goes_to_adjacent_requested_floor():
    outputs = []
    step = '''
    |
    V
'''
    liftA = Lift("A", 0)
    lifts = LiftSystem(floors=[0,1], lifts=[liftA])

    # Initial state
    outputs.append(print_lifts(lifts))

    # Request floor
    liftA.request_floor(1)
    outputs.append(print_lifts(lifts))

    # Move one floor
    lifts.tick()
    outputs.append(print_lifts(lifts))

    # Open doors
    lifts.tick()
    outputs.append(print_lifts(lifts))

    scenario_output = step.join(outputs)
    verify(scenario_output)
