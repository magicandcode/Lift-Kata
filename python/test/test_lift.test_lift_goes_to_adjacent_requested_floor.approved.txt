1         1
0     [A] 0

    |
    V
1      *  1
0     [A] 0

    |
    V
1    [*A] 1
0         0

    |
    V
1     ]A[ 1
0         0
